### PR TITLE
Testing Your Ruby Code - Enumerable!

### PR DESCRIPTION
**Pull request by lmaldonadoch**
On this project, we learned how to use RSpec. We configured it and tested our Enumerable methods with several examples.
## Project milestones
- Install RSpec and create the corresponding files in the project.
- Create examples and test them against our Enumerable methods.
- Use existing repositories for Tic Tac Toe and Enumerables: add specs and open new Pull Requests.
- Add specs for all Enumerables methods you previously created.
- Wrap specs for each method in describe block, hints.
- Instead of using before_each to define a variable for the test cases, use let.
- Use one it block for each expected behaviour. Do not test multiple aspects in one it block.
- Use your common sense: test positive and negative scenarios.

image

## General requirements:
- [x] Add a descriptive README file to your repositories as described here.
- [x] Use meaningful names for your Pull Requests and add a short summary of its content as described here.
- [x] Use descriptive commit messages as described here.
- [x] Do not commit files that are not related to the project. Instead, add them to .gitignore file.
- [x] Do not commit inline comments. They will make your project look messy. If you need to review a previous version of your code, you can always use git history.
## Git Flow:
- [x] Use a specific branch for each feature.
- [x] Use a descriptive name for your branches.
- [x] Use Rubocop to validate your code.

### CODE REVIEW

File: spec/enumerables_spec.rb [hunk 1]
```diff
@@ -0,0 +1,129 @@
+require_relative '../enumerables.rb'
+
+describe Enumerable do
+  let(:array) { [1, 2, 3, 4, 5, 6, 7, 8, 9] }
+  let(:words_array) { %w[ant bear cat] }
+  let(:nil_true) { [nil, true, 99, false] }
+  let(:proced) { proc { |n| n * 2 } }
+  describe '#my_each' do
+    it 'Returns each element' do
+      expect(array.my_each { |n| n * n }).to eql(array)
```
Comment (shshamim63) [inline spec/enumerables_spec.rb]: It actually returns the array itself. I think you can make the test more solid by comparing it with the actual each. I did it in the below format. let me know if you don't like it.
```ruby
it 'Returns the array itself as the actual each' do
      expect(array.my_each { |n| n * n }).to eq(array.each { |n| n * n })
end
```

File: spec/enumerables_spec.rb [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    end
+
+    it 'Returns enumerator' do
+      expect(array.my_each.class).to be(Enumerator)
+    end
+  end
```
Comment (shshamim63) [inline spec/enumerables_spec.rb]: You need to add a more test to check that it actually can call the given block. Check out this one,
```ruby
it 'calls the given block for each of the items in the array' do
      each_output = 0
      block = proc { |num| each_output += num }
      array.my_each(&block)
      my_each_output = each_output.dup
      each_output = 0
      array.each(&block)
      expect(my_each_output).to eq(each_output)
 end

```

File: spec/enumerables_spec.rb [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+  describe '#my_each_with_index' do
+    it 'Returns each element' do
+      expect(array.my_each_with_index { |n| n * n }).to eql(array)
+    end
```
Comment (shshamim63) [inline spec/enumerables_spec.rb]: Same as above, you can compare it with the actual method.

File: spec/enumerables_spec.rb [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+    it 'Returns enumerator' do
+      expect(array.my_each_with_index.class).to be(Enumerator)
+    end
+  end
```
Comment (shshamim63) [inline spec/enumerables_spec.rb]: You can also add a test to check whether it calls the given block or not by adding the following test,
```ruby
it 'calls the given block for each of the items in the array with its index' do
      each_output = 0
      block = proc { |val, idx| each_output += (val + idx) }
      array.my_each_with_index(&block)
      my_each_output = each_output.dup
      each_output = 0
      array.each_with_index(&block)
      expect(my_each_output).to eq(each_output)
end
```

File: spec/enumerables_spec.rb [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+  describe '#my_select' do
+    it 'Returns even numbers' do
+      expect(array.select(&:even?)).to eql([2, 4, 6, 8])
```
Comment (shshamim63) [inline spec/enumerables_spec.rb]: It should be my_select, isn't it?

File: spec/enumerables_spec.rb [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    end
+
+    it 'Returns words shorter than 4 characters' do
+      expect(words_array.select { |n| n.length < 4 }).to eql(%w[ant cat])
```
Comment (shshamim63) [inline spec/enumerables_spec.rb]: It should be my_select too.

File: spec/enumerables_spec.rb [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    end
+
+    it 'Return true for words smaller than 5' do
+      expect(words_array.my_all? { |n| n.length < 5 }).to eql(true)
+    end
+
+    it 'Return false for all equal' do
+      expect(nil_true.my_all?).to eql(false)
+    end
+  end
```
Comment (shshamim63) [inline spec/enumerables_spec.rb]: Add the following test cases to make it complete
- When a class is given as a parameter
- When a regex is given
- When a value is given as a parameter

File: spec/enumerables_spec.rb [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+    end
+
+    it 'Returns false for words shorter than 3 characters' do
+      expect(words_array.my_any? { |n| n.length < 3 }).to eql(false)
+    end
+
+    it 'Returns true for any element in an array' do
+      expect(nil_true.my_any?).to eql(true)
+    end
+  end
```
Comment (shshamim63) [inline spec/enumerables_spec.rb]: Add the following test cases to make it complete

When a class is given as a parameter
When a regex is given
When a value is given as a parameter

File: spec/enumerables_spec.rb [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+    end
+
+    it 'Returns true for words of length 5' do
+      expect(words_array.my_none? { |n| n.length == 5 }).to eql(true)
+    end
+
+    it 'Returns true for empty arrays' do
+      expect([].my_none?).to eql(true)
+    end
+  end
```
Comment (shshamim63) [inline spec/enumerables_spec.rb]: Add the following test cases to make it complete

When a class is given as a parameter
When a regex is given
When a value is given as a parameter

File: spec/enumerables_spec.rb [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+    end
+
+    it 'Returns 1 for true values' do
+      expect(nil_true.my_count { |n| n == true }).to eql(1)
+    end
+  end
+
+  describe '#my_map' do
+    it 'Returns the numbers squared' do
+      expect(array.my_map { |i| i * i }).to eql([1, 4, 9, 16, 25, 36, 49, 64, 81])
```
Comment (shshamim63) [inline spec/enumerables_spec.rb]: [Optional]
Instead of comparing it with an array, it is better to compare it with the actual map.

**Review by shshamim63 — CHANGES_REQUESTED** [review]
## Status :traffic_light: Required Changes :tractor: 
___
Hi @lmaldonadoch and @ferbaco86 :wave: 
almost near to finish this project. Some crucial changes are needed to cover to declare it as complete. Let's take them down.
___
### :spades: Mandatory Changes :spades: 
___
1. In the readme, I could not find any info related to testing. Can you update the Readme with some test info?
2. You need to update some test cases so that it can reflect the actual task. Check the sample I have given.
3. You forgot to add some edge cases test to my_all?, my_none?, my_any?. So please add them.
___
### :clubs: Optional Changes :clubs: 
___
It is always better to compare your method with the actual one. Cause it makes it easy to catch the bugs.
___
##### Please fix the above-mentioned issue and send another review request.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain
[Linkedin](https://www.linkedin.com/in/shakhawathossainshamim/)

File: spec/enumerables_spec.rb (hunk 1 above)
Comment (lmaldonadoch) [reply spec/enumerables_spec.rb]: Done! Thank you for pointing it out! We changed it as suggested

File: spec/enumerables_spec.rb (hunk 2 above)
Comment (lmaldonadoch) [reply spec/enumerables_spec.rb]: Done! Thank you for pointing it out! We added the test case!

File: spec/enumerables_spec.rb (hunk 3 above)
Comment (lmaldonadoch) [reply spec/enumerables_spec.rb]: Done! Thank you for pointing it out! We changed it as suggested

File: spec/enumerables_spec.rb (hunk 4 above)
Comment (lmaldonadoch) [reply spec/enumerables_spec.rb]: Done! Thank you for pointing it out! We added the test case!

File: spec/enumerables_spec.rb (hunk 5 above)
Comment (lmaldonadoch) [reply spec/enumerables_spec.rb]: Done! Thank you for pointing it out! Of course, we had that wrong.

File: spec/enumerables_spec.rb (hunk 6 above)
Comment (lmaldonadoch) [reply spec/enumerables_spec.rb]: Done! Thank you for pointing it out! Of course, we had that wrong.

File: spec/enumerables_spec.rb (hunk 7 above)
Comment (lmaldonadoch) [reply spec/enumerables_spec.rb]: Done! Thank you for pointing it out! We added all the corner case tests for the method!

File: spec/enumerables_spec.rb (hunk 8 above)
Comment (lmaldonadoch) [reply spec/enumerables_spec.rb]: Done! Thank you for pointing it out! We added all the corner case tests for the method!

File: spec/enumerables_spec.rb (hunk 9 above)
Comment (lmaldonadoch) [reply spec/enumerables_spec.rb]: Done! Thank you for pointing it out! We added all the corner case tests for the method!

File: spec/enumerables_spec.rb (hunk 10 above)
Comment (lmaldonadoch) [reply spec/enumerables_spec.rb]: Done! Thank you for pointing it out! We modified all the tests to return exactly the same as the original method!

**Comment by ferbaco86** [comment]
Hello!;
Thanks for the detailed review it has been really helpful. 
We managed to fix each one of the cases commented on your explanation, and also added the required information on the README file about the tests that are being performed. 

Happy coding!

**Review by Torres-ssf — APPROVED** [review]
Project Approved :+1: :clap: :clap: 
Hi there team :hand: 
Great job implementing the requested changes. Your README file looks really nice :clap: 
Good luck with the next project

Happy coding :computer: 
**Sergio Torres**
